File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/bone_mass_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show BoneMassRecord, HealthRecordId, sinceV2_2_0;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers/measurement_unit_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show BoneMassRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [BoneMassRecord] to [BoneMassRecordDto].
@sinceV2_2_0
@internal
extension BoneMassRecordToDto on BoneMassRecord {
  BoneMassRecordDto toDto() {
    return BoneMassRecordDto(
      id: id.toDto(),
      time: time.millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      mass: mass.toDto(),
    );
  }
}

/// Converts [BoneMassRecordDto] to [BoneMassRecord].
@sinceV2_2_0
@internal
extension BoneMassRecordDtoToDomain on BoneMassRecordDto {
  BoneMassRecord toDomain() {
    return BoneMassRecord(
      id: id?.toDomain() ?? HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      mass: mass.toDomain(),
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/BoneMassRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.BoneMassRecord
import androidx.health.connect.client.units.Mass
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.BoneMassRecordDto
import java.time.Instant
import java.time.ZoneOffset

internal fun BoneMassRecord.toDto(): BoneMassRecordDto = BoneMassRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    kilograms = mass.inKilograms,
)

internal fun BoneMassRecordDto.toHealthConnect(): BoneMassRecord = BoneMassRecord(
    mass = Mass.kilograms(kilograms),
    time = Instant.ofEpochMilli(time),
    zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
    metadata = metadata.toHealthConnect(id),
)
